File: api/app/models/menu.py
"""Menu: categories, products, and product option groups/items (modifiers)."""
from __future__ import annotations

import uuid
from decimal import Decimal
from typing import TYPE_CHECKING

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, UUIDMixin

if TYPE_CHECKING:
    from app.models.business import Business


class Category(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "categories"

    business_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("businesses.id", ondelete="CASCADE"), index=True, nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    business: Mapped["Business"] = relationship(back_populates="categories")
    products: Mapped[list["Product"]] = relationship(back_populates="category")


class Product(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "products"
    __table_args__ = (CheckConstraint("price >= 0", name="product_price_nonneg"),)

    business_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("businesses.id", ondelete="CASCADE"), index=True, nullable=False
    )
    category_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("categories.id", ondelete="SET NULL"), index=True
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    price: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    image_url: Mapped[str | None] = mapped_column(String(512))
    is_available: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    is_archived: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    tags: Mapped[list[str]] = mapped_column(ARRAY(String), default=list, nullable=False)
    prep_minutes: Mapped[int | None] = mapped_column(Integer)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)

    business: Mapped["Business"] = relationship(back_populates="products")
    category: Mapped["Category | None"] = relationship(back_populates="products")
    modifier_groups: Mapped[list["ProductModifierGroup"]] = relationship(
        back_populates="product",
        cascade="all, delete-orphan",
        order_by="ProductModifierGroup.sort_order",
    )


class ModifierGroup(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "modifier_groups"
    __table_args__ = (
        CheckConstraint("select_type in ('single','multi')", name="modifier_group_select_type"),
    )

    business_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("businesses.id", ondelete="CASCADE"), index=True, nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    display_name: Mapped[str] = mapped_column(String(255), nullable=False)
    select_type: Mapped[str] = mapped_column(String(16), default="single", nullable=False)
    is_template: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    business: Mapped["Business"] = relationship(back_populates="modifier_groups")
    items: Mapped[list["ModifierOption"]] = relationship(
        back_populates="group",
        cascade="all, delete-orphan",
        order_by="ModifierOption.sort_order",
    )
    assignments: Mapped[list["ProductModifierGroup"]] = relationship(
        back_populates="group", cascade="all, delete-orphan"
    )


class ModifierOption(UUIDMixin, Base):
    __tablename__ = "modifier_options"

    group_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("modifier_groups.id", ondelete="CASCADE"), index=True, nullable=False
    )
    business_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("businesses.id", ondelete="CASCADE"), index=True, nullable=False
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    price_delta: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    is_default: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)

    group: Mapped["ModifierGroup"] = relationship(back_populates="items")


class EffectiveModifierOption:
    def __init__(
        self,
        option: ModifierOption,
        price_delta: Decimal,
        price_delta_override: Decimal | None,
        is_default: bool,
    ) -> None:
        self.id = option.id
        self.name = option.name
        self.description = option.description
        self.price_delta = price_delta
        self.price_delta_override = price_delta_override
        self.is_default = is_default
        self.sort_order = option.sort_order


class ProductModifierGroup(UUIDMixin, Base):
    __tablename__ = "product_modifier_groups"
    __table_args__ = (
        CheckConstraint("min_select >= 0", name="product_modifier_min_nonneg"),
        CheckConstraint(
            "max_select is null or max_select >= 0",
            name="product_modifier_max_nonneg",
        ),
        UniqueConstraint(
            "product_id", "modifier_group_id", name="product_modifier_group_once"
        ),
    )

    business_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("businesses.id", ondelete="CASCADE"), index=True, nullable=False
    )
    product_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("products.id", ondelete="CASCADE"), index=True, nullable=False
    )
    modifier_group_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("modifier_groups.id", ondelete="CASCADE"), index=True, nullable=False
    )
    is_required: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    min_select: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    max_select: Mapped[int | None] = mapped_column(Integer)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)

    product: Mapped["Product"] = relationship(back_populates="modifier_groups")
    group: Mapped["ModifierGroup"] = relationship(back_populates="assignments")
    option_prices: Mapped[list["ProductModifierOptionPrice"]] = relationship(
        back_populates="assignment",
        cascade="all, delete-orphan",
    )

    @property
    def assignment_id(self) -> uuid.UUID:
        return self.id

    @property
    def name(self) -> str:
        return self.group.name

    @property
    def display_name(self) -> str:
        return self.group.display_name

    @property
    def is_template(self) -> bool:
        return self.group.is_template

    @property
    def select_type(self) -> str:
        return self.group.select_type

    @property
    def items(self) -> list[EffectiveModifierOption]:
        options = {item.id: item for item in self.group.items}
        effective: list[EffectiveModifierOption] = []
        for config in self.option_prices:
            option = options.get(config.modifier_option_id)
            if option is None:
                continue
            effective.append(
                EffectiveModifierOption(
                    option,
                    config.price_delta
                    if config.price_delta is not None
                    else option.price_delta,
                    config.price_delta,
                    config.is_default,
                )
            )
        return sorted(effective, key=lambda item: item.sort_order)

    def price_delta_for(self, option_id: uuid.UUID) -> Decimal:
        options = {item.id: item for item in self.group.items}
        for config in self.option_prices:
            if config.modifier_option_id == option_id:
                if config.price_delta is not None:
                    return config.price_delta
                option = options.get(option_id)
                return option.price_delta if option is not None else Decimal("0")
        return Decimal("0")


class ProductModifierOptionPrice(UUIDMixin, Base):
    __tablename__ = "product_modifier_option_prices"
    __table_args__ = (
        UniqueConstraint(
            "product_modifier_group_id",
            "modifier_option_id",
            name="product_modifier_option_price_once",
        ),
    )

    business_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("businesses.id", ondelete="CASCADE"), index=True, nullable=False
    )
    product_modifier_group_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("product_modifier_groups.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )
    modifier_option_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("modifier_options.id", ondelete="CASCADE"),
        index=True,
        nullable=False,
    )
    price_delta: Mapped[Decimal | None] = mapped_column(Numeric(12, 2))
    is_default: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    assignment: Mapped["ProductModifierGroup"] = relationship(
        back_populates="option_prices"
    )
    option: Mapped["ModifierOption"] = relationship()

File: web/src/lib/menu.ts
import { apiUrl } from "@/lib/api";
import { ApiError } from "@/lib/auth";

type Fetcher = typeof fetch;

export type CategoryOut = {
  id: string;
  business_id: string;
  name: string;
  sort_order: number;
  is_active: boolean;
  created_at: string;
};

export type ModifierOptionOut = {
  id: string;
  name: string;
  description: string | null;
  price_delta: string | number;
  is_default: boolean;
  sort_order: number;
};

export type ModifierGroupOut = {
  id: string;
  business_id: string;
  name: string;
  display_name: string;
  select_type: "single" | "multi";
  is_template: boolean;
  items: ModifierOptionOut[];
};

export type ProductModifierGroupOut = {
  assignment_id: string;
  modifier_group_id: string;
  name: string;
  display_name: string;
  select_type: "single" | "multi";
  is_template: boolean;
  is_required: boolean;
  min_select: number;
  max_select: number | null;
  sort_order: number;
  items: (ModifierOptionOut & { price_delta_override: string | number | null })[];
};

export type ProductOut = {
  id: string;
  business_id: string;
  category_id: string | null;
  name: string;
  description: string | null;
  price: string | number;
  image_url: string | null;
  is_available: boolean;
  is_archived: boolean;
  tags: string[];
  prep_minutes: number | null;
  sort_order: number;
  modifier_groups: ProductModifierGroupOut[];
};

export type MenuOptionItemDraft = {
  clientId: string;
  optionId: string | null;
  name: string;
  description: string;
  priceDelta: string;
  priceDeltaOverride: string | null;
  isDefault: boolean;
  isEnabled: boolean;
};

export type MenuOptionGroupDraft = {
  clientId: string;
  modifierGroupId: string | null;
  name: string;
  displayName: string;
  selectType: "single" | "multi";
  isTemplate: boolean;
  isDefinitionEditable: boolean;
  isRequired: boolean;
  minSelect: string;
  maxSelect: string;
  items: MenuOptionItemDraft[];
};

export type MenuDishDraft = {
  id: string | null;
  name: string;
  categoryId: string;
  description: string;
  price: string;
  imageUrl: string;
  isAvailable: boolean;
  tags: string[];
  prepMinutes: string;
  optionGroups: MenuOptionGroupDraft[];
};

export type MenuDishErrors = {
  name?: string;
  categoryId?: string;
  price?: string;
  prepMinutes?: string;
  optionGroups?: Record<string, string>;
};

export type ProductPayload = {
  name: string;
  category_id: string | null;
  description: string | null;
  price: string;
  image_url: string | null;
  is_available: boolean;
  tags: string[];
  prep_minutes: number | null;
  modifier_groups: {
    modifier_group_id?: string;
    definition?: ModifierGroupPayload;
    is_required: boolean;
    min_select: number;
    max_select: number | null;
    sort_order: number;
    items?: {
      option_id: string;
      price_delta: string | null;
      is_default: boolean;
    }[];
  }[];
};

export type ModifierGroupPayload = {
  name: string;
  display_name: string;
  select_type: "single" | "multi";
  is_template: boolean;
  items: {
    id?: string;
    name: string;
    description: string | null;
    price_delta: string;
    is_default: boolean;
    sort_order: number;
  }[];
};

type ProductImageUploadOut = {
  image_url: string;
};

function clientId(prefix: string): string {
  return `${prefix}-${globalThis.crypto?.randomUUID?.() ?? Math.random().toString(36).slice(2)}`;
}

function trimmedOrNull(value: string): string | null {
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : null;
}

function isNonNegativeNumber(value: string, integerOnly = false): boolean {
  if (value.trim() === "") return false;
  const parsed = Number(value);
  return (
    Number.isFinite(parsed) &&
    parsed >= 0 &&
    (!integerOnly || Number.isInteger(parsed))
  );
}

function optionalNumber(value: string): number | null {
  return value.trim() ? Number(value) : null;
}

function moneyOrZero(value: string): string {
  return value.trim() || "0";
}

function stringValue(value: string | number | null): string {
  return value === null ? "" : String(value);
}

export function createOptionItemDraft(
  patch: Partial<MenuOptionItemDraft> = {},
): MenuOptionItemDraft {
  return {
    clientId: clientId("item"),
    optionId: null,
    name: "",
    description: "",
    priceDelta: "0",
    priceDeltaOverride: null,
    isDefault: false,
    isEnabled: true,
    ...patch,
  };
}

export function createOptionGroupDraft(
  patch: Partial<MenuOptionGroupDraft> = {},
): MenuOptionGroupDraft {
  return {
    clientId: clientId("group"),
    modifierGroupId: null,
    name: "",
    displayName: "",
    selectType: "single",
    isTemplate: false,
    isDefinitionEditable: true,
    isRequired: false,
    minSelect: "0",
    maxSelect: "1",
    items: [createOptionItemDraft()],
    ...patch,
  };
}

export function createOptionGroupDraftFromLibrary(
  group: ModifierGroupOut,
): MenuOptionGroupDraft {
  const hasDefaultOption = group.items.some((item) => item.is_default);
  return createOptionGroupDraft({
    modifierGroupId: group.id,
    name: group.name,
    displayName: group.display_name,
    selectType: group.select_type,
    isTemplate: true,
    isDefinitionEditable: false,
    isRequired: hasDefaultOption,
    minSelect: hasDefaultOption ? "1" : "0",
    maxSelect:
      group.select_type === "single" ? "1" : String(group.items.length),
    items: group.items.map((item) =>
      createOptionItemDraft({
        clientId: item.id,
        optionId: item.id,
        name: item.name,
        description: item.description ?? "",
        priceDelta: stringValue(item.price_delta),
        priceDeltaOverride: null,
        isDefault: item.is_default,
        isEnabled: true,
      }),
    ),
  });
}

export function createEmptyDishDraft(categoryId = ""): MenuDishDraft {
  return {
    id: null,
    name: "",
    categoryId,
    description: "",
    price: "",
    imageUrl: "",
    isAvailable: true,
    tags: [],
    prepMinutes: "",
    optionGroups: [],
  };
}

export function createDishDraftFromProduct(
  product: ProductOut,
  modifierGroups: ModifierGroupOut[] = [],
): MenuDishDraft {
  return {
    id: product.id,
    name: product.name,
    categoryId: product.category_id ?? "",
    description: product.description ?? "",
    price: stringValue(product.price),
    imageUrl: product.image_url ?? "",
    isAvailable: product.is_available,
    tags: product.tags,
    prepMinutes: product.prep_minutes === null ? "" : String(product.prep_minutes),
    optionGroups: product.modifier_groups.map((group) => {
      const libraryGroup = modifierGroups.find(
        (candidate) => candidate.id === group.modifier_group_id,
      );
      const enabledById = new Map(group.items.map((item) => [item.id, item]));
      const definitions = libraryGroup?.items ?? group.items;
      const minimum = group.is_required
        ? Math.max(group.min_select, 1)
        : group.min_select;
      const maximum =
        group.select_type === "single"
          ? 1
          : group.max_select && group.max_select > 0
            ? group.max_select
            : enabledById.size;
      return {
        clientId: group.assignment_id,
        modifierGroupId: group.modifier_group_id,
        name: group.name,
        displayName: group.display_name,
        selectType: group.select_type,
        isTemplate: group.is_template,
        isDefinitionEditable: !group.is_template,
        isRequired: group.is_required,
        minSelect: String(minimum),
        maxSelect: String(maximum),
        items: definitions.map((definition) => {
          const enabled = enabledById.get(definition.id);
          return {
            clientId: definition.id,
            optionId: definition.id,
            name: definition.name,
            description: definition.description ?? "",
            priceDelta: stringValue(enabled?.price_delta ?? definition.price_delta),
            priceDeltaOverride: enabled
              ? stringValue(enabled.price_delta_override) || null
              : null,
            isDefault: enabled?.is_default ?? false,
            isEnabled: Boolean(enabled),
          };
        }),
      };
    }),
  };
}

export function validateMenuDishDraft(draft: MenuDishDraft): MenuDishErrors {
  const errors: MenuDishErrors = {};
  const groupErrors: Record<string, string> = {};

  if (!draft.name.trim()) errors.name = "Dish name is required.";
  if (!draft.categoryId) errors.categoryId = "Choose a category.";
  if (!isNonNegativeNumber(draft.price)) {
    errors.price = "Enter a valid non-negative price.";
  }
  if (draft.prepMinutes.trim() && !isNonNegativeNumber(draft.prepMinutes, true)) {
    errors.prepMinutes = "Enter prep time in whole minutes.";
  }

  for (const group of draft.optionGroups) {
    if (!group.name.trim()) {
      groupErrors[group.clientId] = "Group name is required.";
      continue;
    }
    if (!group.displayName.trim()) {
      groupErrors[group.clientId] = "Customer label is required.";
      continue;
    }
    const enabledItems = group.items.filter((item) => item.isEnabled);
    if (enabledItems.length === 0) {
      groupErrors[group.clientId] = "Enable at least one option.";
      continue;
    }
    if (
      group.selectType === "multi" &&
      (!group.maxSelect.trim() ||
        !isNonNegativeNumber(group.maxSelect, true) ||
        Number(group.maxSelect) < 1)
    ) {
      groupErrors[group.clientId] = "Maximum choices must be a positive whole number.";
      continue;
    }
    const invalidItem = group.items.find(
      (item) =>
        !item.name.trim() || !isNonNegativeNumber(item.priceDelta || "0"),
    );
    if (invalidItem) {
      groupErrors[group.clientId] =
        "Each option item needs a name and non-negative price delta.";
    }
  }

  if (Object.keys(groupErrors).length > 0) errors.optionGroups = groupErrors;
  return errors;
}

export function hasMenuDishErrors(errors: MenuDishErrors): boolean {
  return Boolean(
    errors.name ||
      errors.categoryId ||
      errors.price ||
      errors.prepMinutes ||
      (errors.optionGroups && Object.keys(errors.optionGroups).length > 0),
  );
}

export function buildModifierGroupPayload(
  group: MenuOptionGroupDraft,
): ModifierGroupPayload {
  return {
    name: group.name.trim(),
    display_name: group.displayName.trim(),
    select_type: group.selectType,
    is_template: group.isTemplate,
    items: group.items.map((item, itemIndex) => ({
      ...(item.optionId ? { id: item.optionId } : {}),
      name: item.name.trim(),
      description: (item.description ?? "").trim() || null,
      price_delta: moneyOrZero(item.priceDelta),
      is_default: item.isDefault,
      sort_order: itemIndex,
    })),
  };
}

export function buildProductPayload(
  draft: MenuDishDraft,
): ProductPayload {
  return {
    name: draft.name.trim(),
    category_id: draft.categoryId || null,
    description: trimmedOrNull(draft.description),
    price: moneyOrZero(draft.price),
    image_url: trimmedOrNull(draft.imageUrl),
    is_available: draft.isAvailable,
    tags: draft.tags.map((tag) => tag.trim()).filter(Boolean),
    prep_minutes: optionalNumber(draft.prepMinutes),
    modifier_groups: draft.optionGroups.map((group, groupIndex) => {
      const enabledCount = group.items.filter((item) => item.isEnabled).length;
      const maximumChoices =
        group.selectType === "single"
          ? 1
          : Number(group.maxSelect) > 0
            ? Number(group.maxSelect)
            : enabledCount;
      return {
        ...(group.modifierGroupId
          ? { modifier_group_id: group.modifierGroupId }
          : {}),
        ...(group.isDefinitionEditable
          ? { definition: buildModifierGroupPayload(group) }
          : {}),
        is_required: group.isRequired,
        min_select: group.isRequired ? 1 : 0,
        max_select: maximumChoices,
        sort_order: groupIndex,
        items: group.isTemplate && !group.isDefinitionEditable
          ? group.items
              .filter((item) => item.isEnabled && item.optionId)
              .map((item) => ({
                option_id: item.optionId as string,
                price_delta: item.priceDeltaOverride,
                is_default: item.isDefault,
              }))
          : undefined,
      };
    }),
  };
}

async function menuRequest<T>({
  path,
  accessToken,
  method = "GET",
  body,
  fetcher = fetch,
}: {
  path: string;
  accessToken: string;
  method?: "GET" | "POST" | "PATCH" | "DELETE";
  body?: unknown;
  fetcher?: Fetcher;
}): Promise<T> {
  let response: Response;
  try {
    response = await fetcher(apiUrl(path), {
      method,
      headers: {
        Authorization: `Bearer ${accessToken}`,
        ...(body === undefined ? {} : { "Content-Type": "application/json" }),
      },
      ...(body === undefined ? {} : { body: JSON.stringify(body) }),
    });
  } catch {
    throw new ApiError(
      "Couldn't reach the server. Is the API running on http://localhost:8000?",
      0,
    );
  }

  if (!response.ok) {
    let message = "Couldn't save your menu. Please try again.";
    let code: string | undefined;
    try {
      const payload = (await response.json()) as {
        error?: { message?: string; code?: string };
      };
      message = payload.error?.message ?? message;
      code = payload.error?.code;
    } catch {
      // Keep the stable fallback for non-JSON responses.
    }
    throw new ApiError(message, response.status, code);
  }

  if (response.status === 204) return undefined as T;
  return (await response.json()) as T;
}

export function listCategories({
  accessToken,
  businessId,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  fetcher?: Fetcher;
}): Promise<CategoryOut[]> {
  return menuRequest<CategoryOut[]>({
    path: `/businesses/${businessId}/categories`,
    accessToken,
    fetcher,
  });
}

export function createCategory({
  accessToken,
  businessId,
  name,
  sortOrder = 0,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  name: string;
  sortOrder?: number;
  fetcher?: Fetcher;
}): Promise<CategoryOut> {
  return menuRequest<CategoryOut>({
    path: `/businesses/${businessId}/categories`,
    accessToken,
    method: "POST",
    body: { name: name.trim(), sort_order: sortOrder, is_active: true },
    fetcher,
  });
}

export function updateCategoryActive({
  accessToken,
  businessId,
  categoryId,
  isActive,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  categoryId: string;
  isActive: boolean;
  fetcher?: Fetcher;
}): Promise<CategoryOut> {
  return menuRequest<CategoryOut>({
    path: `/businesses/${businessId}/categories/${categoryId}`,
    accessToken,
    method: "PATCH",
    body: { is_active: isActive },
    fetcher,
  });
}

export function updateCategoryName({
  accessToken,
  businessId,
  categoryId,
  name,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  categoryId: string;
  name: string;
  fetcher?: Fetcher;
}): Promise<CategoryOut> {
  return menuRequest<CategoryOut>({
    path: `/businesses/${businessId}/categories/${categoryId}`,
    accessToken,
    method: "PATCH",
    body: { name: name.trim() },
    fetcher,
  });
}

export function deleteCategory({
  accessToken,
  businessId,
  categoryId,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  categoryId: string;
  fetcher?: Fetcher;
}): Promise<void> {
  return menuRequest<void>({
    path: `/businesses/${businessId}/categories/${categoryId}`,
    accessToken,
    method: "DELETE",
    fetcher,
  });
}

export function listModifierGroups({
  accessToken,
  businessId,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  fetcher?: Fetcher;
}): Promise<ModifierGroupOut[]> {
  return menuRequest<ModifierGroupOut[]>({
    path: `/businesses/${businessId}/modifier-groups`,
    accessToken,
    fetcher,
  });
}

export function createModifierGroup({
  accessToken,
  businessId,
  group,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  group: MenuOptionGroupDraft;
  fetcher?: Fetcher;
}): Promise<ModifierGroupOut> {
  return menuRequest<ModifierGroupOut>({
    path: `/businesses/${businessId}/modifier-groups`,
    accessToken,
    method: "POST",
    body: buildModifierGroupPayload(group),
    fetcher,
  });
}

export function updateModifierGroup({
  accessToken,
  businessId,
  groupId,
  group,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  groupId: string;
  group: MenuOptionGroupDraft;
  fetcher?: Fetcher;
}): Promise<ModifierGroupOut> {
  return menuRequest<ModifierGroupOut>({
    path: `/businesses/${businessId}/modifier-groups/${groupId}`,
    accessToken,
    method: "PATCH",
    body: buildModifierGroupPayload(group),
    fetcher,
  });
}

export function deleteModifierGroup({
  accessToken,
  businessId,
  groupId,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  groupId: string;
  fetcher?: Fetcher;
}): Promise<void> {
  return menuRequest<void>({
    path: `/businesses/${businessId}/modifier-groups/${groupId}`,
    accessToken,
    method: "DELETE",
    fetcher,
  });
}

export function listProducts({
  accessToken,
  businessId,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  fetcher?: Fetcher;
}): Promise<ProductOut[]> {
  return menuRequest<ProductOut[]>({
    path: `/businesses/${businessId}/products`,
    accessToken,
    fetcher,
  });
}

export async function uploadMenuItemImage({
  accessToken,
  businessId,
  image,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  image: Blob;
  fetcher?: Fetcher;
}): Promise<string> {
  let response: Response;
  try {
    response = await fetcher(
      apiUrl(`/businesses/${businessId}/products/images`),
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${accessToken}`,
          "Content-Type": image.type || "application/octet-stream",
        },
        body: image,
      },
    );
  } catch {
    throw new ApiError(
      "Couldn't reach the server. Is the API running on http://localhost:8000?",
      0,
    );
  }

  if (!response.ok) {
    let message = "Couldn't upload the image. Please try again.";
    let code: string | undefined;
    try {
      const payload = (await response.json()) as {
        error?: { message?: string; code?: string };
      };
      message = payload.error?.message ?? message;
      code = payload.error?.code;
    } catch {
      // Keep the stable fallback for non-JSON responses.
    }
    throw new ApiError(message, response.status, code);
  }

  const payload = (await response.json()) as ProductImageUploadOut;
  return payload.image_url;
}

export async function saveMenuDish({
  accessToken,
  businessId,
  draft,
  categories,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  draft: MenuDishDraft;
  categories: CategoryOut[];
  fetcher?: Fetcher;
}): Promise<ProductOut> {
  const errors = validateMenuDishDraft(draft);
  if (hasMenuDishErrors(errors)) {
    throw new ApiError("Check your dish details before saving.", 422);
  }
  if (!categories.some((category) => category.id === draft.categoryId)) {
    throw new ApiError("Choose a category that belongs to this business.", 422);
  }

  return menuRequest<ProductOut>({
    path: draft.id
      ? `/businesses/${businessId}/products/${draft.id}`
      : `/businesses/${businessId}/products`,
    accessToken,
    method: draft.id ? "PATCH" : "POST",
    body: buildProductPayload(draft),
    fetcher,
  });
}

export function updateProductAvailability({
  accessToken,
  businessId,
  productId,
  isAvailable,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  productId: string;
  isAvailable: boolean;
  fetcher?: Fetcher;
}): Promise<ProductOut> {
  return menuRequest<ProductOut>({
    path: `/businesses/${businessId}/products/${productId}`,
    accessToken,
    method: "PATCH",
    body: { is_available: isAvailable },
    fetcher,
  });
}

export function deleteProduct({
  accessToken,
  businessId,
  productId,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  productId: string;
  fetcher?: Fetcher;
}): Promise<void> {
  return menuRequest<void>({
    path: `/businesses/${businessId}/products/${productId}`,
    accessToken,
    method: "DELETE",
    fetcher,
  });
}

export async function seedSampleMenu({
  accessToken,
  businessId,
  categories,
  products,
  fetcher = fetch,
}: {
  accessToken: string;
  businessId: string;
  categories: CategoryOut[];
  products: ProductOut[];
  fetcher?: Fetcher;
}): Promise<{ categories: CategoryOut[]; products: ProductOut[] }> {
  const desiredCategories = ["Starters", "Mains", "Drinks", "Desserts"];
  const allCategories = [...categories];

  for (const [index, name] of desiredCategories.entries()) {
    if (!allCategories.some((category) => category.name === name)) {
      allCategories.push(
        await createCategory({
          accessToken,
          businessId,
          name,
          sortOrder: index,
          fetcher,
        }),
      );
    }
  }

  const mains = allCategories.find((category) => category.name === "Mains");
  if (!mains) return { categories: allCategories, products };

  const seededProducts = [...products];
  const sampleProducts = [
    {
      name: "Classic Beef Burger",
      description: "Smash patty, cheddar, house sauce",
      price: "850",
      tags: ["bestseller", "halal"],
      prepMinutes: "20",
      optionGroups: [
        createOptionGroupDraft({
          clientId: "seed-size",
          name: "Burger sizes",
          displayName: "Size",
          selectType: "single",
          isRequired: true,
          minSelect: "1",
          maxSelect: "1",
          items: [
            createOptionItemDraft({
              clientId: "seed-regular",
              name: "Regular",
              priceDelta: "0",
              isDefault: true,
            }),
            createOptionItemDraft({
              clientId: "seed-large",
              name: "Large",
              priceDelta: "250",
            }),
          ],
        }),
        createOptionGroupDraft({
          clientId: "seed-addons",
          name: "Burger add-ons",
          displayName: "Add-ons",
          selectType: "multi",
          minSelect: "0",
          maxSelect: "3",
          items: [
            createOptionItemDraft({
              clientId: "seed-cheese",
              name: "Extra cheese",
              priceDelta: "150",
            }),
            createOptionItemDraft({
              clientId: "seed-jalapenos",
              name: "Jalapenos",
              priceDelta: "80",
            }),
            createOptionItemDraft({
              clientId: "seed-no-onions",
              name: "No onions",
              priceDelta: "0",
            }),
          ],
        }),
      ],
    },
    {
      name: "Creamy Alfredo Pasta",
      description: "Fettuccine, parmesan, herbs",
      price: "900",
      tags: ["veg"],
      prepMinutes: "25",
      optionGroups: [],
    },
  ];

  for (const sample of sampleProducts) {
    if (seededProducts.some((product) => product.name === sample.name)) continue;
    seededProducts.push(
      await saveMenuDish({
        accessToken,
        businessId,
        categories: allCategories,
        draft: {
          ...createEmptyDishDraft(mains.id),
          name: sample.name,
          description: sample.description,
          price: sample.price,
          tags: sample.tags,
          prepMinutes: sample.prepMinutes,
          optionGroups: sample.optionGroups,
        },
        fetcher,
      }),
    );
  }

  return { categories: allCategories, products: seededProducts };
}
